esrally/mechanic/telemetry.py

```diff
@@ -481,107 +481,90 @@ def record(self):
             if self.include_buffer_pools:
-                self.record_jvm_buffer_pool_stats(node_name, node_stats)
+                collected_node_stats.update(self.collect_jvm_buffer_pool_stats(node_name, node_stats))
             if self.include_mem_stats:
-                self.record_jvm_mem_stats(node_name, node_stats)
+                collected_node_stats.update(self.collect_jvm_mem_stats(node_name, node_stats))
             if self.include_network:
-                self.record_network_stats(node_name, node_stats)
+                collected_node_stats.update(self.collect_network_stats(node_name, node_stats))
             if self.include_process:
-                self.record_process_stats(node_name, node_stats)
+                collected_node_stats.update(self.collect_process_stats(node_name, node_stats))
+
+            self.metrics_store.put_doc(collected_node_stats,
+                                       level=MetaInfoScope.node,
+                                       node_name=node_name,
+                                       meta_data=self.metrics_store_meta_data)
 
         time.sleep(self.sample_interval)
 
-    def record_indices_stats(self, node_name, node_stats, include):
+    def flatten_stats_fields(self, prefix=None, stats=None, include=None, level=0):
+        """
+        Flatten provided dict using an optional prefix and top level key filters.
+
+        :param prefix: The prefix for all flattened values. _ will be used as a separator. Defaults to None.
+        :param stats: Dict with values to be flattened, using _ as a separator. Defaults to {}.
+        :param include: List of keys to process in stats. Filtering will be done only for top level keys.
+        :param level: Int used to control filtering at the top level; only used internally. Defaults to 0.
+        :return:
+        """
+
+        def iterate():
+            for section_name, section_value in stats.items():
+                # Process only top level matching keys
+                if include is None or level==0 and section_name in include:
+                    if isinstance(section_value, dict):
+                        for subsection_name, subsection_value in self.flatten_stats_fields(stats=section_value, level=level+1).items():
+                            # Avoid duplication for metric fields that have unit embedded in value as they are also recorded elsewhere
+                            # example: `breakers_parent_limit_size_in_bytes` vs `breakers_parent_limit_size`
+                            if isinstance(subsection_value, (int, float)) and not isinstance(subsection_value, bool):
+                                yield "{}{}_{}".format(prefix + "_" if prefix else "", section_name, subsection_name), subsection_value
+                    else:
+                        if isinstance(section_value, (int, float)) and not isinstance(section_value, bool):
```

Comment:
I wonder whether we can get rid of this code duplication when simplifying this method as suggested above.